### PR TITLE
Colossus Implementation for GCP

### LINKED ISSUE #45: Colossus: Titan at Ludicrous Scale
"Colossus" will be Titan at the largest scale that will fit into a /16. GCP will be created first as its implementation will be the most simple of the clouds.

Each layer will span a /19 and a /20 to provide 12,288 hosts per layer. This will accommodate massive Kubernetes deployments. Almost the entire /16 space will be occupied by these five layers.

### PR DESCRIPTION
**Pull request by naftulikay**
Closes #45.

### CODE REVIEW

**Comment by naftulikay** [comment]
As of now, we have public networking working, private networking working, zonal NAT working, and regional NAT working.

Might be a candidate for a merge, only thing left is to revise I/O on the NAT module.

**Comment by naftulikay** [comment]
A few thoughts on NAT:

### Stop Viewing NAT Gateways as Servers

Gateways are stateless and are terminated when malfunctioning and replaced.

The only configuration variables that are relevant to NAT:

1. `max_availability_zones` to limit the amount of gateways.
2. `zonal_high_availability` to make NAT highly available for each zone.
3. `instance_type` to increase bandwidth by increasing CPU

[Google says:](https://cloud.google.com/vpc/docs/special-configurations?authuser=0#multiple-natgateways) 

> Bandwidth per gateway is up to 2 Gbps per core unidirectional.

### Try Using `next_hop_instance`

In Google's example, `next_hop_instance` and `next_hop_instance_zone` are used rather than `next_hop_ip`:

```shell
gcloud compute routes create natroute1 --destination-range 0.0.0.0/0 \
    --tags no-ip --priority 800 --next-hop-instance-zone us-east1-b \
    --next-hop-instance $nat_1_instance
```

This makes everything with `no-ip` route through the NAT Gateway in the same zone as the instance.

There are a couple questions here:
 - [ ] If an instance is destroyed and recreated, will the route be invalidated and need recreation?
 - [ ] If one AZ is down, will instances in that AZ fail over to the next AZ?

Regardless, there should be two NAT route priorities:

 1. `850`: general, cross-zone NAT via the tags `nat` and `private`
 2. `800`: local zone NAT with the tags `nat-${az}`, `nat`, and `private`

Testing should prove out if `800` fails that the `850` rule is matched.

### Add Dependency Ordering for Zonal HA NAT

In order to limit downtime during NAT changes, make the secondary resources depend on the primary resources. This should ensure that the primary is provisioned and brought back into service before continuing onto the secondary.

---

### TODO

#### Zonal NAT

 - [ ] Create a new `800` route priority for `next_hop_instance` and `next_hop_instance_zone`.
 - [ ] With the above configuration, kill an instance and determine if repair happens automatically.
 - [ ] With the above configuration, set all MIGs in a zone to size of 0 to see if the next route is matched.

#### NAT Dependencies

 - [ ] Make all secondary NAT resources depend on the primaries.

#### NAT Cleanup

 - [ ] Clean up I/O of the NAT module to only accept as few variables as possible, view NAT gateways as hosted services.

**Comment by naftulikay** [comment]
### Zonal NAT

In response to the last comment, we cannot easily proceed with `next_hop_instance` and `next_hop_instance_zone` to provide zonal default NAT routes. We can't do this because of the following limitations:
 1. In order to get an instance from a `google_compute_instance_group_manager`, you need to define a `data` resource to reference the underlying group.
 2. You need the data resource to execute at the _end_ of everything so that you are guaranteed your instances are present.
 3. In order to make this happen, Google uses a module for MIGs which has an `output` for `instances` which references the data source.
 4. Thus, we would need to define a module for each AZ which creates the MIG and returns a list of its instances.
 5. Terraform does not allow us to have a `count` on modules.
 6. Therefore, we can't proceed with the current design, we would have to hard-code a fixed set of modules into the `colossus_nat` module, which we're not going to do.

**TL;DR** It's _really_ hard to setup zonal NAT following Google's example, due to Terraform module/count issues that have been around for years.

#### Regional/Zonal NAT Data

Additionally, I captured a series of traceroutes from the `simple` instance in the Admin Layer as defined in the attached example. [The data can be seen in full here](https://gist.github.com/naftulikay/92e2bbb08fec83735fe8027d30b999c2).

The average ping time from the instance to an arbitrary NAT gateway is _0.651 milliseconds_, namely 651 _microseconds_. This is averaged across six NAT gateway instances across three separate AZs (two NAT gateways in each zone). This is really not bad. Let's do some statistics via Python 3 over the data:

##### Regional

 - `statistics.mean`: 0.651ms
 - `statistics.median`: 0.635ms
 - `statistics.mode` in Python freaks out if there are more than 1 most common value; there are 7.
 - `statistics.pstdev`: Population Standard Deviation: 0.190ms
 - `statistics.pvariance`: Population Variance: 0.036ms
 - `statistics.stdev`: Standard Deviation: 0.191ms

We're talking about a standard deviation of around 200 _microseconds_ across three availability zones. ECMP seems to be doing its trick.

##### Zonal

In-zone-only NAT actually seems to take _more time_ than regional NAT, presumably because there are less hosts to service a flow:

 - `statistics.mean`: 0.878ms
 - `statistics.median`: 0.651ms (0.6505, rounded)
 - `statistics.mode`: 5 equally common values
 - `statistics.pstdev`: Population Standard Deviation: 0.449ms
 - `statistics.pvariance`: Population Variance: 0.202ms
 - `statistics.stev`: Standard Deviation: 0.452ms

Thus: **regionally-distributed NAT is substantially faster than zone-only NAT**.

In conclusion, I'm going to rip out zone-only NAT for now, based on the data I am seeing.

**Comment by naftulikay** [comment]
It also appears that changing the instance type at runtime still works, flows take longer as they are retried, but it works. This means that upgrading NAT is possible provided that you have it setup for zonal HA so that primary instances can be rotated while secondary instances are online, then proceed to secondary instances while primaries are online.

**Comment by naftulikay** [comment]
In summary:
 - Public networking works.
 - Private networking works.
 - GKE works.
 - Sharding instances across AZs (similar to AWS) works.
 - Regional instance groups (similar to AWS ASG) work.
 - NAT is HA across all zones (for 3 AZs, 3 NAT gateway routes)
 - NAT is optionally HA within a zone (`nat_zonal_high_availability` enables two NAT gateways per zone)

For private Kubernetes clusters, you need a `/28`. Use something like `172.16.0.16/28`, stolen [from Google's private GKE tutorial](https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters). Try to use ranges within the private `172.16.0.0/12` block as detailed in RFC-1819.
